SRCH-4366: Upgrade i14y to Rails ~> 7.0.x AND SRCH-3883: Remove Ruby 2.7 from i14y CircleCI

luisgmetzger opened the pull request:
## Summary
- This PR upgrades the current Rails version to 7.0
- This should follow most of the required items that needed to be checked and changed in the Rails upgrade documentation, found here: [Rails Upgrade Documentation](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-6-1-to-rails-7-0)
- Here is Jira ticket link: https://cm-jira.usa.gov/browse/SRCH-4366
- This also updates the CircleCI config file to remove the check for ruby version 2.7.5
 
### Checklist
Please ensure you have addressed all concerns below before marking a PR "ready for review" or before requesting a re-review. If you cannot complete an item below, replace the checkbox with the ⚠️ `:warning:` emoji and explain why the step was not completed.
 
#### Functionality Checks
 
- [x] You have merged the latest changes from the target branch (usually `main`) into your branch.
  
- [x] Your primary commit message is of the format **SRCH-#### \<description\>** matching the associated Jira ticket.

- [x] PR title is either of the format **SRCH-#### \<description\>** matching the associated Jira ticket (i.e. "SRCH-123 implement feature X"), or **Release - SRCH-####, SRCH-####, SRCH-####** matching the Jira ticket numbers in the release.
 
- [x] Automated checks pass. If Code Climate checks do not pass, explain reason for failures:
 
#### Process Checks

- [x] You have specified at least one "Reviewer".

lorawoodford commented on spec/classes/document_search_spec.rb:
```diff
@@ -199,14 +199,13 @@ def create_documents(document_hashes)
-        expect(NewRelic::Agent).to receive(:notice_error).with(
-          error, options: { custom_params: { indices: ['test-i14y-documents-agency_blogs'] } }
-        )
+        allow(NewRelic::Agent).to receive(:notice_error).and_return(nil)
```
Same as above... move the `allow` up into the before, then we can change this to:

```
     document_search.search
     
     expect(NewRelic::Agent).to have_received(:notice_error)
```

luisgmetzger commented:
@lorawoodford - I have addressed your comments.

lorawoodford commented on Gemfile.lock:
```diff
@@ -177,31 +184,31 @@ GEM
-    nokogiri (1.15.3)
-      mini_portile2 (~> 2.8.2)
+    nio4r (2.7.0)
+    nokogiri (1.15.5-arm64-darwin)
```
We don't want to be committing platform-specific changes... I'm not sure why this is happening, but I can't replicate locally.  You may want to check `bundle config`, confirm you're using a ruby version manager, and/or do some additional digging to get to the bottom of this.  This is also what's causing the unwanted updates to `Platforms` below.

luisgmetzger replied:
@lorawoodford - This makes sense. In previous projects where we use Docker environments, this doesn't come up much. I can change this.

lorawoodford commented on Gemfile.lock:
```diff
@@ -347,10 +353,11 @@ GEM
 
 PLATFORMS
-  ruby
+  arm64-darwin-23
```
This is likely to cause a proliferation of other developer-specific platforms here and seems an unnecessary/unintended change?  `ruby` is sufficient and what we have in every other repo.

lorawoodford reviewed (changes requested):
Sorry @luisgmetzger but one additional issue you'll need to dig into that seems to be particular to your current environment.  Happy to help troubleshoot or feel free to post something in the slack devs channel to get additional eyes on this if needed.

luisgmetzger commented:
@lorawoodford  - I have addressed your comments. I have updated the Gemfile.lock to only use the Ruy platform. I checked the dependencies stated in this file to confirm it is not using a stated platform. Looks to be good now.

lorawoodford reviewed (approved):
LGTM!
